File: 124-binary-tree-maximum-path-sum/124-binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def maxPathSum(self, root):
        node = root
        # global res
        self.res = node.val
        
        def dfs(node):
            global res
            if not node:
                return 0 
            leftmax = dfs(node.left)
            rightmax = dfs(node.right)
            
            leftmax = max(leftmax, 0)
            rightmax = max(rightmax, 0)
            self.res = max(self.res, (node.val + leftmax + rightmax))
            
            return(max((node.val + leftmax),(node.val + rightmax)))
        dfs(node)
        return(self.res)
        
        
        
        
        """
        :type root: TreeNode
        :rtype: int
        """
        

File: 124-binary-tree-maximum-path-sum/124-binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int max;
    public int maxPathSum(TreeNode root) {
        max = Integer.MIN_VALUE;
        recursion(root);
        return max;
    }
    
    public int recursion(TreeNode root) {
        if(root==null) {
            return 0;
        }
        int left = Math.max(recursion(root.left),0);
        int right = Math.max(recursion(root.right),0);
        
        max = Math.max(left+right+root.val,max);
        
        return Math.max(left,right)+root.val;
    }
}